File: python/parallel.py
from mpi4py import MPI
import sys
import time
import math
import copy
from util import create_2matrix, add_and_multiply, print_result
from save_matrix import SaveMatrix

if __name__ == '__main__':
    comm = MPI.COMM_WORLD
    p = comm.Get_size()
    rank = comm.Get_rank()

    p_sqrt = int(math.sqrt(p-1))
    m_size = int(sys.argv[1])
    p_size = int(m_size / p_sqrt)
    if rank == 0:
        matrix1, matrix2 = create_2matrix(m_size)
        save_matrix = SaveMatrix(matrix1, matrix2, p-1)
        start_time = time.time()
        matrix = [[0] * m_size for _ in range(m_size)]
        for i in range(p_sqrt):
            for j in range(p_sqrt):
                dest = p_sqrt * i + j + 1
                row = p_size * i
                col = p_size * j
                mtx1 = []
                mtx2 = []
                for r in range(row, row + p_size):
                    mtx1.append([])
                    mtx2.append([])
                    for c in range(col, col + p_size):
                        mtx1[r % p_size].append(matrix1[r][(c + r) % m_size])
                        mtx2[r % p_size].append(matrix2[(r + c) % m_size][c])
                mtx3 = [matrix[r][j * p_size:j * p_size + p_size] for r in range(i * p_size, i * p_size + p_size)]
                data = [mtx1, mtx2, mtx3]
                # data = [mtx1, mtx2, [[0] * p_size for i in range(p_size)]]
                comm.send(data, dest=dest, tag=dest)

        for i in range(p_sqrt):
            for j in range(p_sqrt):
                dest = p_sqrt * i + j + 1
                mtx = comm.recv(source=dest, tag=dest)
                for k in range(p_size):
                    matrix[i * p_size + k][j * p_size: j * p_size + p_size] = mtx[m_size-1][k]
                save_matrix.add_to_c_parallel(mtx, dest)

        end_time = time.time()
        print_result(matrix, end_time - start_time)
        save_matrix.to_json_file()
    else:
        data = comm.recv(source=0, tag=rank)
        r_one = rank - 1
        r_size = rank - p_sqrt
        x = rank + 1
        y = rank + p_sqrt
        p_sqrt_2 = p_sqrt ** 2
        dest1 = r_one if r_one % p_sqrt != 0 else r_one + p_sqrt
        dest2 = r_size if r_size > 0 else p_sqrt_2 + r_size
        source1 = x if rank % p_sqrt != 0 else x - p_sqrt
        source2 = y if y <= p_sqrt_2 else y - p_sqrt_2
        init = True
        all_iter = []
        for t in range(m_size):
            add_and_multiply(data[2], data[0], data[1], len(data[0]))
            all_iter.append(copy.deepcopy(data[2]))
            if t == 0:
                init = False
            if t == m_size - 1:
                break
            comm.send([i[0] for i in data[0]], dest=dest1, tag=dest1)
            comm.send(data[1][0], dest=dest2, tag=dest2)
            new_col = comm.recv(source=source1, tag=rank)
            for i in range(p_size):
                data[0][i] = data[0][i][1:] + [new_col[i]]
            new_row = comm.recv(source=source2, tag=rank)
            data[1] = data[1][1:] + [new_row]
        req = comm.send(all_iter, dest=0, tag=rank)

File: golang/parallel/parallel.go
package parallel

import (
	"math"
	"time"

	"github.com/MicaTravica/ntp/golang/savematrix"
	"github.com/MicaTravica/ntp/golang/util"
)

type resultP struct {
	Num    int
	Matrix [][]float64
}

func pMatrix(n int, mtx [][][]float64, chanals map[string]chan []float64, num int, main chan resultP, saveMatrix savematrix.SaveMatrix) {
	pSize := len(mtx[0][0])
	for t := 0; t < n; t++ {
		util.AddAndMultiply(mtx[2], mtx[0], mtx[1], pSize)
		saveMatrix.AddToCParallel(mtx[2], num, t)
		if t == n-1 {
			main <- resultP{num, mtx[2]}
			break
		}
		d1 := make([]float64, pSize)
		for i := 0; i < pSize; i++ {
			d1[i] = mtx[0][i][0]
		}
		chanals["dest1"] <- d1
		chanals["dest2"] <- mtx[1][0]
		s1 := <-chanals["source1"]
		for i := 0; i < pSize; i++ {
			mtx[0][i] = append(mtx[0][i][1:], s1[i])

		}
		s2 := <-chanals["source2"]
		mtx[1] = append(mtx[1][1:], s2)

	}
}

// Parallel matrix multiplication
func Parallel(n, p int) {
	matrix1, matrix2 := util.CreateMatrix2(n)

	saveMatrix := savematrix.SaveMatrix{}
	saveMatrix.CrateSaveMatrix(matrix1, matrix2, p)

	start := time.Now()

	result := make([][]float64, n)
	for i := range result {
		result[i] = make([]float64, n)
		for j := range result[i] {
			result[i][j] = 0
		}
	}

	pSqrt := int(math.Sqrt(float64(p)))
	pSize := int(n / pSqrt)

	chanals := make([][][]chan []float64, pSqrt)
	for i := range chanals {
		chanals[i] = make([][]chan []float64, pSqrt)
		for j := range chanals[i] {
			chanals[i][j] = make([]chan []float64, 2)
			for k := range chanals[i][j] {
				chanals[i][j][k] = make(chan []float64, 1)
			}
		}
	}
	main := make(chan resultP)
	for i := 0; i < pSqrt; i++ {
		for j := 0; j < pSqrt; j++ {
			mtx := make([][][]float64, 3)
			chs := make(map[string]chan []float64)
			mtx[0] = make([][]float64, pSize)
			mtx[1] = make([][]float64, pSize)
			mtx[2] = make([][]float64, pSize)
			row := i * pSize
			col := j * pSize
			for r := row; r < row+pSize; r++ {
				rM := r % pSize
				mtx[0][rM] = make([]float64, pSize)
				mtx[1][rM] = make([]float64, pSize)
				mtx[2][rM] = make([]float64, pSize)
				for c := col; c < col+pSize; c++ {
					cM := c % pSize
					mtx[0][rM][cM] = matrix1[r][(c+r)%n]
					mtx[1][rM][cM] = matrix2[(r+c)%n][c]
				}
			}
			chs["dest1"] = chanals[i][(j-1+pSqrt)%pSqrt][0]
			chs["dest2"] = chanals[(i-1+pSqrt)%pSqrt][j][1]
			chs["source1"] = chanals[i][j][0]
			chs["source2"] = chanals[i][j][1]
			go pMatrix(n, mtx, chs, i*pSqrt+j, main, saveMatrix)
		}
	}
	for t := 0; t < p; t++ {
		rp := <-main
		for k := 0; k < pSize; k++ {
			i := rp.Num/pSqrt*pSize + k
			j := rp.Num % pSqrt * pSize
			result[i] = append(append(result[i][:j], rp.Matrix[k]...), result[i][j+pSize:]...)
		}
	}
	elapsed := time.Since(start)

	util.PrintResult(result, elapsed)
	saveMatrix.WriteToFile()
}
